File: generated_code/GeminiPro/Java/037_SplayTree.java
/**
 * A complete, self-contained implementation of an integer Splay Tree.
 * Splay trees are self-balancing binary search trees with the property that
 * recently accessed elements are quick to access again. This implementation
 * supports insert, delete, and search operations for a set of integers.
 */
public class SplayTree {

    /**
     * Inner class to represent a node in the Splay Tree.
     * It includes a parent pointer to facilitate the splaying operation.
     */
    private static class Node {
        int key;
        Node parent;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.parent = null;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Searches for a key in the tree.
     * If the key is found, the node containing the key is splayed to the root.
     * If the key is not found, the last accessed node before the search failed
     * is splayed to the root.
     *
     * @param key The integer key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        Node current = root;
        Node lastNode = null;

        while (current != null) {
            lastNode = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key found, splay the node and return the key
                splay(current);
                return current.key;
            }
        }

        // Key not found. Splay the last accessed node if the tree was not empty.
        if (lastNode != null) {
            splay(lastNode);
        }
        return null;
    }

    /**
     * Inserts a key into the Splay Tree.
     * If the key already exists, the existing node is splayed to the root.
     * If the key is new, it is inserted and the new node is splayed to the root.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        Node node = root;
        Node parent = null;

        // 1. Find position for new node (standard BST insert)
        while (node != null) {
            parent = node;
            if (key < node.key) {
                node = node.left;
            } else if (key > node.key) {
                node = node.right;
            } else {
                // Key already exists, splay it and return
                splay(node);
                return;
            }
        }

        // 2. Insert the new node
        Node newNode = new Node(key);
        newNode.parent = parent;

        if (parent == null) {
            root = newNode; // Tree was empty
        } else if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }

        // 3. Splay the new node to the root
        splay(newNode);
    }

    /**
     * Deletes a key from the Splay Tree.
     * The tree is first searched for the key, which brings the node (or its
     * would-be parent) to the root. If the key is found at the root, it is
     * removed and the tree is re-joined.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        // Splay the node with the given key (or its parent) to the root
        search(key);

        // If the key is not in the tree, the root will not have the key.
        if (root == null || root.key != key) {
            return; // Key not found, nothing to delete.
        }

        // Now, the node to delete is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left child, the right child becomes the new root.
            root = rightSubtree;
            if (root != null) {
                root.parent = null;
            }
        } else {
            // Detach the left subtree to operate on it independently.
            leftSubtree.parent = null;

            // Find the maximum node in the left subtree.
            Node maxInLeft = leftSubtree;
            while (maxInLeft.right != null) {
                maxInLeft = maxInLeft.right;
            }

            // Splay this maximum node to the top of the left subtree.
            // Temporarily set the main root to the left subtree's root to do this.
            this.root = leftSubtree;
            splay(maxInLeft);

            // After splaying, `this.root` is `maxInLeft`.
            // The new root (`maxInLeft`) has no right child.
            // Attach the original right subtree as its right child.
            this.root.right = rightSubtree;
            if (rightSubtree != null) {
                rightSubtree.parent = this.root;
            }
        }
    }

    /**
     * Performs the splay operation on a node x, moving it to the root.
     * @param x The node to splay.
     */
    private void splay(Node x) {
        if (x == null) return;
        
        while (x.parent != null) {
            Node parent = x.parent;
            Node grandParent = parent.parent;

            if (grandParent == null) { // Zig case
                if (x == parent.left) {
                    rightRotate(parent);
                } else {
                    leftRotate(parent);
                }
            } else {
                if (x == parent.left) {
                    if (parent == grandParent.left) { // Zig-Zig case
                        rightRotate(grandParent);
                        rightRotate(parent);
                    } else { // Zig-Zag case
                        rightRotate(parent);
                        leftRotate(grandParent);
                    }
                } else { // x is the right child
                    if (parent == grandParent.left) { // Zig-Zag case
                        leftRotate(parent);
                        rightRotate(grandParent);
                    } else { // Zig-Zig case
                        leftRotate(grandParent);
                        leftRotate(parent);
                    }
                }
            }
        }
        // The loop terminates when x is the root. The rotation methods handle
        // updating the main root field when the old root is rotated.
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * @param x The node to rotate.
     */
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y != null) {
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.left = x;
            x.parent = y;
        }
    }

    /**
     * Performs a right rotation on the subtree rooted at node x.
     * @param x The node to rotate.
     */
    private void rightRotate(Node x) {
        Node y = x.left;
        if (y != null) {
            x.left = y.right;
            if (y.right != null) {
                y.right.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.right) {
                x.parent.right = y;
            } else {
                x.parent.left = y;
            }
            y.right = x;
            x.parent = y;
        }
    }
}

File: generated_code/GeminiPro/Python/037_SplayTree.py
import sys

# Set a higher recursion limit for deep trees, although this implementation is iterative.
# This is more of a safeguard for other potential recursive operations not present here.
sys.setrecursionlimit(2000)

class SplayTree:
    """
    A self-contained SplayTree class that implements a dictionary-like set for integers.

    A splay tree is a self-adjusting binary search tree with the additional property
    that recently accessed elements are quick to access again. It performs basic
    operations such as insertion, look-up and removal in O(log n) amortized time.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty SplayTree."""
        self.root = None

    def _splay(self, key):
        """
        Performs the splay operation.

        Brings the node with the given key, or the last accessed node on the search path
        if the key is not found, to the root of the tree. This is an iterative,
        top-down splay implementation.
        """
        if not self.root:
            return

        # Dummy node to simplify linking of left and right subtrees
        dummy = self._Node(None)
        left_tree_tail = dummy
        right_tree_tail = dummy

        current = self.root

        while True:
            if key < current.key:
                if not current.left:
                    break
                # Zig-Zig case (right-right)
                if key < current.left.key:
                    # Rotate right
                    temp = current.left
                    current.left = temp.right
                    temp.right = current
                    current = temp
                    if not current.left:
                        break
                # Link current node to the right tree
                right_tree_tail.left = current
                right_tree_tail = current
                current = current.left
            elif key > current.key:
                if not current.right:
                    break
                # Zig-Zig case (left-left)
                if key > current.right.key:
                    # Rotate left
                    temp = current.right
                    current.right = temp.left
                    temp.left = current
                    current = temp
                    if not current.right:
                        break
                # Link current node to the left tree
                left_tree_tail.right = current
                left_tree_tail = current
                current = current.right
            else:  # key == current.key
                break

        # Reassemble the tree
        left_tree_tail.right = current.left
        right_tree_tail.left = current.right
        current.left = dummy.right
        current.right = dummy.left
        self.root = current

    def search(self, key):
        """
        Searches for a key in the tree and splays the accessed node.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise. After the search, the node
            containing the key (if found) or the last accessed node (if not found)
            is moved to the root.
        """
        if not self.root:
            return False
        
        self._splay(key)
        
        # After splaying, the root will be the key if it exists.
        return self.root.key == key

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, it is splayed to the root. If it doesn't exist,
        it is inserted and becomes the new root.

        Args:
            key: The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        self._splay(key)

        # If key is already in the tree, splay has brought it to the root.
        if self.root.key == key:
            return

        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            self.root.left = None
            self.root = new_node
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            self.root.right = None
            self.root = new_node

    def delete(self, key):
        """
        Deletes a key from the tree.

        If the key is found, it is deleted. The tree is re-structured by splaying.
        If the key is not found, the tree structure may change due to the splay
        operation, but no node is removed.

        Args:
            key: The integer key to delete.
        """
        if not self.root:
            return

        self._splay(key)

        # If the key is not in the tree, do nothing after splaying.
        if self.root.key != key:
            return

        # The key to be deleted is now at the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left subtree, the right subtree becomes the new tree.
            self.root = right_subtree
        else:
            # Make the left subtree the main tree temporarily.
            self.root = left_subtree
            # Splay for the key we just deleted (or any key larger than all
            # elements in the left subtree). This brings the largest element
            # of the left subtree to its root.
            self._splay(key)
            # The new root (max element of the original left subtree) now has
            # no right child. We can attach the original right subtree here.
            self.root.right = right_subtree